File: src/lib/std/mango/string.m

struct String {
	val length I32
	val chars Ptr<I16>
}

fn String.toInt (radix Int) Int -> {
	var n = 0
	var p = 1
	val length = this.length
	var i = length - 1
	while i >= 0 {
		n += unsafe { this.chars[i] - '0' } * p
		p *= radix
		i -= 1
	}
	n
}

[inline]
fn String.toInt Int -> this.toInt(10)

fn Int.toString (radix Int) String -> {
    use std.text.builder*

    if this == 0 return "0"

    val builder = StringBuilder {
        length: 0
        chars: Ptr<I16> { length: 10 }
        capacity: 10
    }

    var isNegative = false
    var num = this

    if num < 0 && radix == 10 {
        isNegative = true
        num = -num
    }

    while num != 0 {
        val rem = (num % radix) as I16
        if rem > 9 builder.appendChar((rem - 10) + 'a')
        : builder.appendChar(rem + '0')
        num /= radix
    }

    if isNegative builder.appendChar('-')

    builder.invert()

    builder.toString()
}

[inline]
fn Int.toString String -> this.toString(10)

[inline]
fn Bool.toString String -> if this return "true" : return "false"

[operator]
fn String.equals (other String) Bool -> {
	val size = this.length
	if size != other.length return false
    var i = 0
    while i < size {
        if this[i] != other[i] return false
        i += 1
    }
	true
}

[inline]
[operator]
fn String.get(i I32) I16 -> unsafe { this.chars[i] }

File: src/lib/std/c/string.c

String* intToString (Int num, Int radix) {
	String* string = malloc(sizeof(String));
	string->chars = malloc(50);
	string->length = 0;

    Bool isNegative = false;

    if (num == 0) {
        string->chars[string->length++] = '0';
        return string;
    }

    if (num < 0 && radix == 10) {
        isNegative = true;
        num = -num;
    }

    while (num != 0) {
        Int rem = num % radix;
        string->chars[string->length++] = (rem > 9) ? (rem - 10) + 'a' : rem + '0';
        num /= radix;
    }

    if (isNegative) {
        string->chars[string->length++] = '-';
    }

	Int i = 0;
	Int j = string->length - 1;
	while (i < (string->length / 2 + string->length % 2)) {
		char tmp = string->chars[i];
		string->chars[i] = string->chars[j];
		string->chars[j] = tmp;
		i++;
		j--;
	}

    return string;
}

Int stringToInt (String* string, Int radix) {
	Int n = 0;
	Int p = 1;
	U32 length = string->length;
	for (Int i = length - 1; i >= 0; i--) {
		n += ((Int)(string->chars[i] - '0')) * p;
		p *= radix;
	}
	return n;
}

Bool String$equals (String* str0, String* str1) {
	U32 size = str0->length;
	if (size != str1->length) return false;
    U32 i = 0;
    while (i < size) {
        if (str0->chars[i] != str1->chars[i]) {
            return false;
        }
        i++;
    }
	return true;
}
